internal/api/core/middleware/proxy.go

```diff
@@ -1,118 +1,225 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package middleware
 
 import (
+	"crypto/tls"
 	"fmt"
 	"net"
 	"net/http"
 	"net/http/httputil"
-	"net/url"
 	"regexp"
 	"time"
 
 	"github.com/labstack/echo/v4"
 	"github.com/perses/common/etcd"
 	"github.com/perses/perses/internal/api/interface/v1/datasource"
+	"github.com/perses/perses/internal/api/interface/v1/globaldatasource"
+	v1 "github.com/perses/perses/pkg/model/api/v1"
+	datasourcev1 "github.com/perses/perses/pkg/model/api/v1/datasource"
 	"github.com/sirupsen/logrus"
 )
 
 var (
-	proxyMatcher                       = regexp.MustCompile(`/proxy/datasources/([a-zA-Z-0-9_-]+)(/.*)?`)
-	defaultTransport http.RoundTripper = &http.Transport{
-		Proxy: http.ProxyFromEnvironment,
-		DialContext: (&net.Dialer{
-			Timeout:   30 * time.Second,
-			DualStack: true,
-			KeepAlive: 30 * time.Second,
-		}).DialContext,
-		TLSHandshakeTimeout: 10 * time.Second,
-		MaxIdleConns:        100,
-		IdleConnTimeout:     90 * time.Second,
-		ForceAttemptHTTP2:   true,
-	}
+	globalProxyMatcher = regexp.MustCompile(`/proxy/globaldatasources/([a-zA-Z-0-9_-]+)(/.*)?`)
+	localProxyMatcher  = regexp.MustCompile(`/proxy/projects/([a-zA-Z-0-9_-]+)/datasources/([a-zA-Z-0-9_-]+)(/.*)?`)
 )
 
-func Proxy(dao datasource.DAO) echo.MiddlewareFunc {
+func Proxy(dts datasource.DAO, globalDTS globaldatasource.DAO) echo.MiddlewareFunc {
 	return func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
-			targetURL, err := buildProxy(c, dao)
+			spec, path, err := catchDatasourceAndPath(c, dts, globalDTS)
 			if err != nil {
 				return err
 			}
-			if targetURL == nil {
+			if spec == nil {
 				return next(c)
 			}
-			return serveProxy(c, targetURL)
+			pr, err := newProxy(spec, path)
+			if err != nil {
+				return err
+			}
+			return pr.serve(c)
 		}
 	}
 }
 
-func buildProxy(c echo.Context, dao datasource.DAO) (*url.URL, error) {
-	req := c.Request()
-	requestPath := req.URL.Path
-	if !proxyMatcher.MatchString(requestPath) {
-		// request is not matching the proxy path. Probably it's a request to the API itself
-		// That's why no error is returned in order to let the request passed.
-		return nil, nil
-	}
-	logrus.Tracef("'%s' is a request to a datasource", requestPath)
-	matchingGroups := proxyMatcher.FindAllStringSubmatch(requestPath, -1)
+func catchDatasourceAndPath(c echo.Context, dts datasource.DAO, globalDTS globaldatasource.DAO) (v1.DatasourceSpec, string, error) {
```

Comment:
`catch` sounds a bit unusual for this kind of operation, how about `extract`?

Reply:
extract is definitely better :). Thanks !